File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stck;
        int i;
        char c;
        
        if(s.length()==0)
          return true;
        if(s.length()==1)
          return false;
        for(i=0;i<s.length();i++)
        {
            if((s[i]=='(') || (s[i]=='{') || (s[i]=='['))
                stck.push(s[i]);
            else 
            {
                if(stck.size()==0)
                  return false;
                switch(s[i])
                {
                  case ')':
                    if(stck.top()!='(')
                      return false;
                    break;
                    
                  case '}':
                    if(stck.top()!='{')
                      return false;
                    break;
                    
                  case ']':
                    if(stck.top()!='[')
                      return false;
                    break;
                }
              
                stck.pop();
            }
        }
            
        if(!stck.empty())
            return false;
        else
          return true;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')':'(', ']':'[', '}':'{'}
        for c in s:
            if c in mapping:
                if stack:
                    top = stack.pop()
                    if top!=mapping[c]:
                        return False
                else:
                    return False
            else:
                stack.append(c)
        if stack:
            return False
        return True
        #O(N), O(N)

